File: leetcode/[0053]maximum-subarray.ts
// https://leetcode.com/problems/maximum-subarray/
function maxSubArray(nums: number[]): number {
    return solution3(nums);
}



// brute force: time limit exceeded
function solution1(nums: number[]): number {
    let maxs = Number.MIN_SAFE_INTEGER;

    for (let i = 0; i < nums.length; i++) {
        let subtotal = 0;
        for (let j = i; j < nums.length; j++) {
            subtotal += nums[j];
            maxs = Math.max(maxs, subtotal);
        }
    }

    return maxs;
}



// kadane's algorithm: 4 ms
function solution2(nums: number[]): number {
    const maxs = Array(nums.length).fill(0);
    maxs[0] = nums[0];

    for (let i = 1; i < nums.length; i++) {
        maxs[i] = Math.max(maxs[i - 1] + nums[i], nums[i]);
    }

    return Math.max(...maxs);
}



// divide and conquer: 10 ms
function solution3(nums: number[]): number {
    return dac(nums)[1];
}
function dac(nums: number[]): [number, number, number, number] {
    if (nums.length === 1) return [nums[0], nums[0], nums[0], nums[0]];

    const m = Math.trunc(nums.length / 2);
    const [ll, lm, lr, lt] = dac(nums.slice(0, m));
    const [rl, rm, rr, rt] = dac(nums.slice(m,));

    return [
        Math.max(ll, lt + rl),
        Math.max(lm, lr + rl, rm),
        Math.max(lr + rt, rr),
        lt + rt
    ];
}

File: leetcode/[0053]maximum-subarray.rs
// https://leetcode.com/problems/maximum-subarray/
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        return solution3(nums);
    }
}



// brute force: time limit exceeded
fn solution1(nums: Vec<i32>) -> i32 {
    use std::cmp::max;
    
    let mut max_sub = i32::MIN;

    for i in 0..nums.len() {
        let mut sub_total = 0;
        for j in i..nums.len() {
            sub_total += nums[j];

            max_sub = max(max_sub, sub_total);
        }
    }

    return max_sub;
}



// kadane's algo: 3 ms
fn solution2(nums: Vec<i32>) -> i32 {
    use std::cmp::max;
    
    let mut max_subs = vec![0; nums.len()];
    max_subs[0] = nums[0];
    let mut max_sub = nums[0];

    for (i, &x) in (1..).zip(nums[1..].iter()) {
        max_subs[i] = max(max_subs[i - 1] + x, x);
        max_sub = max(max_sub, max_subs[i]);
    }

    return max_sub;
}



// divide and conquer: 3 ms
fn solution3(nums: Vec<i32>) -> i32 {
    return dac(&nums)[1];
}
fn dac(nums: &[i32]) -> [i32; 4] {
    use std::cmp::max;

    if nums.len() == 1 {
        return [nums[0], nums[0], nums[0], nums[0]];
    }

    let m = nums.len() / 2;
    let [ll, lm, lr, lt] = dac(&nums[..m]);
    let [rl, rm, rr, rt] = dac(&nums[m..]);

    return [max(ll, lt + rl), max(lm, max(lr + rl, rm)), max(lr + rt, rr), lt + rt];
}
